cpp/src/arrow/dataset/filter.cc

```diff
@@ -1261,7 +1264,232 @@ Result<std::shared_ptr<RecordBatch>> TreeEvaluator::Filter(
   return batch->Slice(0, 0);
 }
 
-std::shared_ptr<ScalarExpression> scalar(bool value) { return scalar(MakeScalar(value)); }
+std::shared_ptr<Expression> scalar(bool value) { return scalar(MakeScalar(value)); }
+
+// Serialization is accomplished by converting expressions to single element StructArrays
+// then writing that to an IPC file. The last field is always an int32 column containing
+// ExpressionType, the rest store the Expression's members.
```

Comment:
Seems likely we'll want to define a (Flatbuffers?) serialization protocol protocol for expressions in the not-too-distant future, though this certainly works for now! =)